### PR TITLE
feature(ui-ux): `pool swap` - store last selected slippage tolerance

### LINKED ISSUE #2031: `pool swap` - store last selected slippage tolerance
#### What would you like to be added:
On the pool swap page, we can select or specify a different slippage tolerance. With the new design, it takes two steps to change it. We would like to store the last slippage tolerance value that the user entered so they don't need to modify it on future transactions.

image

- [ ] When user selects a Slippage tolerance (custom or not), it will be stored on their device
- [ ] Future pool swaps will use this new slippage tolerance as a default value

### PR DESCRIPTION
**Pull request by chloezxyy**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
To store the last slippage tolerance value that the user entered so they don't need to modify it on future transactions.

#### Which issue(s) does this PR fixes?:

Fixes #2031

#### Developer Checklist:

- [x] Read your code changes at least once
- [x] Tested on iOS/Android device (e.g, No crashes, library supported etc.)
- [x] No console errors on web
- [ ] Tested on Light mode and Dark mode*
- [ ] Your UI implementation visually matched the rendered design*
- [ ] Unit tests*
- [x] Added e2e tests*
- [ ] Added translations*

### CODE REVIEW

File: mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts [hunk 1]
```diff
@@ -149,6 +149,10 @@ context('Wallet - DEX - Composite Swap with balance', () => {
 
     cy.getByTestID('slippage_input').clear().type('25').blur().wait(100)
     cy.getByTestID('button_tolerance_submit').click()
+
+    cy.url().should('include', 'app/DEX/CompositeSwap', () => {
+      expect(localStorage.getItem('WALLET.SLIPPAGE_TOLERANCE')).to.eq('5')
+    })
```
Comment (kyleleow) [inline mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts:155]: Lets separate the tests:
- check if able to store new slippage
- check default slippage

**Review by kyleleow — CHANGES_REQUESTED** [review]

File: mobile-app/app/screens/AppNavigator/screens/Dex/CompositeSwap/CompositeSwapScreen.tsx [hunk 2]
```diff
@@ -491,7 +492,9 @@ export function CompositeSwapScreen ({ route }: Props): JSX.Element {
               fee={fee}
               isConversionRequired={isConversionRequired}
               slippage={slippage}
-              onSetSlippage={(val: BigNumber) => setSlippage(val)}
+              onSetSlippage={(val: BigNumber) => {
+                void setSlippage(val)
```
Comment (kyleleow) [inline mobile-app/app/screens/AppNavigator/screens/Dex/CompositeSwap/CompositeSwapScreen.tsx:496]: any reason to add `void` here?

File: mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { SlippageTolerancePersistence } from '@api'
+import { useLogger } from '@shared-contexts/NativeLoggingProvider'
+import BigNumber from 'bignumber.js'
+import { useEffect, useState } from 'react'
+
+export interface SlippageTolerance {
```
Comment (kyleleow) [inline mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts:6]: we are using this interface somewhere else? if not can remove `export`?

File: mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts [hunk 4]
```diff
@@ -126,6 +126,9 @@ context('Wallet - DEX - Composite Swap with balance', () => {
     cy.getByTestID('text_input_tokenA').type('10')
     cy.getByTestID('slippage_select').click()
     cy.getByTestID('slippage_1%').should('exist')
+    cy.url().should('include', 'app/DEX/CompositeSwap', () => {
```
Comment (kyleleow) [inline mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts:129]: is the `cy.url()..` necessary? similarly for the two below

File: mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts [hunk 5]
```diff
@@ -150,6 +153,18 @@ context('Wallet - DEX - Composite Swap with balance', () => {
...
   })
+
+  it('should be able to store selected slippage value in storage', () => {
+    cy.url().should('include', 'app/DEX/CompositeSwap', () => {
+      expect(localStorage.getItem('WALLET.SLIPPAGE_TOLERANCE')).to.eq('25')
+    })
+  })
+
+  it('previously saved slippage tolerance value should be 25%', () => {
+    cy.getByTestID('text_input_tokenA').type('10')
```
Comment (kyleleow) [inline mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts:164]: use a different amount, based on previous steps, user already typed `10` before reaching this step

File: mobile-app/app/api/persistence/slippage-tolerance_storage.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import AsyncStorage from '@react-native-async-storage/async-storage'
+import BigNumber from 'bignumber.js'
+
+const KEY = 'WALLET.SLIPPAGE_TOLERANCE'
+
+async function set (slippageTolerance: NonNullable<BigNumber>): Promise<void> {
+  await AsyncStorage.setItem(KEY, slippageTolerance.toString())
```
Comment (kyleleow) [inline mobile-app/app/api/persistence/slippage-tolerance_storage.ts:7]:
```suggestion
  await AsyncStorage.setItem(KEY, slippageTolerance.toFixed(8))
```

File: mobile-app/app/api/persistence/slippage-tolerance_storage.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+}
+async function get (): Promise<BigNumber> {
```
Comment (kyleleow) [inline mobile-app/app/api/persistence/slippage-tolerance_storage.ts:9]: how do you return a `BigNumber` when you're only storing `string`?

File: mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    slippage: BigNumber
+    setSlippage: (val: BigNumber) => Promise<void>
+}
+
+export function useSlippageTolerance (): SlippageTolerance {
+    const logger = useLogger()
+    const [slippage, setSlippage] = useState(new BigNumber(1))
+
+    useEffect(() => {
+        SlippageTolerancePersistence.get().then((slippageValue: BigNumber) => {
```
Comment (kyleleow) [inline mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts:16]:
```suggestion
        SlippageTolerancePersistence.get().then((slippageValue: string) => {
```

File: mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+          setSlippage(new BigNumber(slippageValue))
+        }).catch(logger.error)
+    }, [])
+
+    const updateSlippageTolerance = async (slippageVal: BigNumber): Promise<void> => {
+        setSlippage(new BigNumber(slippageVal))
```
Comment (kyleleow) [inline mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts:22]: if your argument is already of type BigNumber, do you still need `new BigNumber()`?

**Review by kyleleow — CHANGES_REQUESTED** [review]

File: mobile-app/app/screens/AppNavigator/screens/Dex/CompositeSwap/CompositeSwapScreen.tsx (hunk 2 above)
Comment (chloezxyy) [reply mobile-app/app/screens/AppNavigator/screens/Dex/CompositeSwap/CompositeSwapScreen.tsx:496]: setSlippage is a Promise, so it must be marked with a void

File: mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts (hunk 3 above)
Comment (chloezxyy) [reply mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts:6]: This is the only interface that we have

File: mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts (hunk 4 above)
Comment (chloezxyy) [reply mobile-app/cypress/integration/functional/wallet/dex/compositeswap.spec.ts:130]: As discussed, its a cypress command enqueued operation that is used with async storage usage. Will leave this for now.

**Review by kyleleow — DISMISSED** [review]

File: mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts (hunk 3 above)
Comment (pierregee) [reply mobile-app/app/screens/AppNavigator/screens/Dex/hook/SlippageTolerance.ts:6]: `export` can be removed, I don't see any other hook/component using it

**Review by pierregee — APPROVED** [review]

**Review by thedoublejay — APPROVED** [review]

**Pull request merged**